### ISSUE TITLE
[Enhancement] Ability to remove the timestamp from external link(s)

### ISSUE THREAD

**Issue by alondrayona** [iv-org/invidious#4197]
**Is your enhancement request related to a problem? Please describe.**

After [#4101](https://github.com/iv-org/invidious/pull/4101) the watch-on-youtube link is automatically updated with the timestamp. I don't like this because I'm using this link to share a video, but I usually don't want to add the timestamp.
I think it would be better to create a separate element (or action) to create the timestamped link than to add it to the external link by default.

**Describe the solution you'd like**

The default external link should not include the timestamp.

**Describe alternatives you've considered**

- Create a config option to control this behavior
- Create custom link (or action) to timestamped link

**Additional context**

**Comment by Tha14** [comment]
Yes, please add an option for this in the settings. I tend to either copy and share the youtube link to friends or play it on an external player and it is an extra step that becomes annoying quite quickly. Cheers

**Comment by gsuberland** [comment]
I don't know Crystal but I was able to kludge together an implementation of this. If someone who actually knows the language wants to properly review this and turn it into a PR, that'd be helpful. The main thing I'm unsure on is the preferences logic in `views/watch.ecr`, as I don't know what happens if the preference is unset (is it called `nil` in Crystal?) there, if it even can be.

Here's a patchdiff for the changes I made: [invidious_no_timestamps_patchdiff.txt](https://github.com/iv-org/invidious/files/13462012/invidious_no_timestamps_patchdiff.txt)

Here are the detailed alterations:

### src/invidious/config.cr

Add `property time_in_external_links : Bool = true` to the `ConfigPreferences` struct.

I made the assumption here that the default should be true to match the current behaviour of including timestamps.

### src/invidious/views/user/preferences.cr

Duplicate the control template for saving player position and change it to:

```
            <div class="pure-control-group">
                <label for="time_in_external_links"><%= translate(locale, "preferences_time_in_external_links_label") %></label>
                <input name="time_in_external_links" id="time_in_external_links" type="checkbox" <% if preferences.time_in_external_links %>checked<% end %>>
            </div>
```

This makes the preference appear in the user preferences page.

### src/invidious/user/preferences.cr

Look for the `save_player_pos` property within the `Preferences` struct and add a new property after it:

```
  property time_in_external_links : Bool = CONFIG.default_user_preferences.time_in_external_links
```

This adds the preference to the data structure which I think is later used to form the JSON data.

### src/invidious/routes/preferences.cr

Search each instance of `save_player_pos` and add the exact same handling for `time_in_external_links`. This involved:

Adding this after the `save_player_pos` handling code:

```
    time_in_external_links = env.params.body["time_in_external_links"]?.try &.as(String)
    time_in_external_links ||= "off"
    time_in_external_links = time_in_external_links == "on"
```

Adding `time_in_external_links: time_in_external_links,` to the end of the `Preferences.from_json({ ... })` call.

### src/invidious/videos/video_preferences.cr

Add `property time_in_external_links : Bool` to the properties in the `VideoPreferences` struct.

In the `process_video_params` function, look for references to `save_player_pos` and duplicate the code for `time_in_external_links`. This involved:

Adding this at the end of the parsing code, after `save_player_pos` gets parsed:

```
  time_in_external_links = query["time_in_external_links"]?.try { |q| (q == "true" || q == "1").to_unsafe }
```

Adding this in the `if preferences` block:

```
    time_in_external_links ||= preferences.time_in_external_links.to_unsafe
```

Adding this at the end of the code that sets default options if none are explicitly set:

```
  time_in_external_links ||= CONFIG.default_user_preferences.time_in_external_links.to_unsafe
```

Adding this at the end of the block where all the config values are normalised from 0/1 to true/false:

```
  time_in_external_links = time_in_external_links == 1
```

And at the very end, adding `time_in_external_links: time_in_external_links,` to the end of the `VideoPreferences.new({ ... }` call.

All of this is to handle the preference being set in the preferences page. I'm used to this being handled in a controller rather than a route but I'm guessing Crystal does things differently here.

### src/invidious/views/watch.ecr

Change this code:

https://github.com/iv-org/invidious/blob/9ce9c543992243737516750bf08f5d073e899715/src/invidious/views/watch.ecr#L120-L124

To this:

```
                    if !plid.nil? && !continuation.nil? && env.get("preferences").as(Preferences).time_in_external_links
                        link_yt_param = URI::Params{"plid" => [plid], "index" => [continuation.to_s]}
                        link_yt_watch = IV::HttpServer::Utils.add_params_to_url(link_yt_watch, link_yt_param)
                        link_yt_embed = IV::HttpServer::Utils.add_params_to_url(link_yt_embed, link_yt_param)
                    end
```

I'm not 100% sure what this code is supposed to do in the first place, but I think it adds a timestamp to the external URLs at page load when you're resuming watching using the "save playback position" setting. If this is the case, my modification prevents the continuation parameters from being added to the external URL on page load. This needs review by someone who actually knows the codebase well.

### assets/js/player.js

The player JS includes an event callback for `timeupdate` which updates the links dynamically during playback.

https://github.com/iv-org/invidious/blob/9ce9c543992243737516750bf08f5d073e899715/assets/js/player.js#L126-L129

We need to stop that callback from being registered when the setting is disabled. This is easily done by wrapping it with an if statement:

```js
/**
 * Callback that updates the timestamp on all external links
 */
if (video_data.preferences.time_in_external_links) {
    player.on('timeupdate', function () {
        ...
    });
}
```

If there's a possibility that `time_in_external_links` will be null here, it probably wants `?? true` on the ends of the condition to default it to including timestamps.

### locales/en-US.json

Add an entry for `preferences_time_in_external_links_label`. I went with "Include timestamp in external links: " as the text for the preference.

This would, of course, need to be duplicated for all the other supported locales, but the only other language I could vaguely take a crack at is German and the results would still probably be pretty bad!
